src/lib/operators/join_hash.cpp

```diff
@@ -687,24 +687,29 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
                              _right_in_table->get_chunk(ChunkID{0})->get_column(ColumnID{0}))
                              ? true
                              : false;
-    auto only_output_right_input = !_inputs_swapped || (_mode != JoinMode::Semi && _mode != JoinMode::Anti);
+    auto only_output_right_input = _inputs_swapped && (_mode == JoinMode::Semi || _mode == JoinMode::Anti);
 
     /**
-     * Two Caches to avoid redundant reference materialization for Reference input tables. As the there might be
+     * Two Caches to avoid redundant reference materialization for Reference input tables. As there might be
      *  quite a lot Partitions (>500 seen), input Chunks (>500 seen), and columns (>50 seen), this speeds up
-     *  write_output_chunks a lot()
+     *  write_output_chunks a lot.
      *
      * They do two things:
      *      - Make it possible to re-use output pos lists if two columns in the input table have exactly the same
      *          PosLists Chunk by Chunk
      *      - Avoid creating the std::vector<const PosList*> for each Partition over and over again.
+     *
+     * They hold one entry per column in the table, not per BaseColumn in a single chunk
      */
     PosListsByColumn left_pos_lists_by_column;
     PosListsByColumn right_pos_lists_by_column;
 
-    if (ref_col_left && only_output_right_input) {
+    // left_pos_lists_by_column won't be used if left is data table or not being output at all
+    if (ref_col_left && !only_output_right_input) {
       left_pos_lists_by_column = setup_pos_lists_by_column(_left_in_table);
     }
+
+    // right_pos_lists_by_column won't be used if right is data table
```

Comment:
Ok, but then why do we set it up? Or in other words: Why do we have that check for the left side but not for the right side?

Reply:
the right side is always written to the output_table

Reply:
So it _will_ be used?

Reply:
right_pos_lists_by_column will NOT be used if the right input is a DATA table only if it is a REFERENCE table.

Reply:
wow, that sounds angry: It's in write_output_columns: the caches won't be used for data tables thus it makes no sense to fill them.

Reply:
I could refactor the JoinHash to have different write_output methods for data and reference inputs but then again, the entire JoinHash could be refactored....

Reply:
Ah, I think I didn't understand the term "data table". We don't use it anywhere in the code (yet).

If you change this to "right_pos_lists_by_column will only be used if right is a reference table", we're good to go.

Reply:
changed